File: leetcode_ly/c++/List/ConvertSortedListToBinarySearchTree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedListToBST(ListNode* head) {
        return buildTree(head,NULL);
    }
    
    TreeNode* buildTree(ListNode* head,ListNode* end){
        if(head == end) return NULL;
        ListNode* low = head;
        ListNode* high = head;
        while(high->next != end && high->next->next != end){
            low = low->next;
            high = high->next->next;
        }
        TreeNode* root = new TreeNode(low->val);
        root->left = buildTree(head,low);
        root->right = buildTree(low->next,end);
        return root;
    }
};

File: leetcode_ly/java/List/ConvertSortedListToBinarySearchTree.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode sortedListToBST(ListNode head) {
        return buildTree(head,null);
    }
    
    public TreeNode buildTree(ListNode head,ListNode end){
        if(head == end) return null;
        ListNode slow = head;
        ListNode fast = head;
        //下面这句话注意！！和一般的两指针不一样，终止条件是!= end
        while(fast.next != end && fast.next.next != end){
            slow = slow.next;
            fast = fast.next.next;
        }
        TreeNode root = new TreeNode(slow.val);
        root.left = buildTree(head,slow);
        root.right = buildTree(slow.next,end);
        return root;
    }
}